Fix decomposition for inverse S and T gates

josh146 opened the pull request:
**Context:**

Two bugs were recently discovered in #955.

* The tape QNode was decomposing gates based on the contents of `device.operations`. Instead, it should have been using `device.supports_operation()`, which has additional logic to deal with inverses.

* The new `Operation.expand()` method which is used in tape mode was not inverting the decomposed tape if the original op was inverted.

**Description of the Change:**

* The tape QNode expand logic has been changed to use `device.supports_operation()`.

* The `Operation.expand()` method now inverts the expanded tape if the operation was inverted.

* Adds tests to make sure this logic works.

**Benefits:**

* Devices that don't natively support `qml.S.inv()` and `qml.T.inv()` are now provided with the correct decomposition.

* In tape mode, devices that *do* support `qml.S.inv()` and `qml.T.inv()` do not have these operations decomposed.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** Fixes #995

glassnotes commented on pennylane/ops/qubit.py:
```diff
@@ -257,10 +257,10 @@ def _matrix(cls, *params):
-        return decomp_ops
+    def decomposition(self, wires):
+        if self.inverse:
+            return [PhaseShift(-np.pi / 2, wires=wires)]
```
:100:

glassnotes reviewed (approved):
Thanks for taking care of this so quickly @josh146 !

antalszava reviewed (changes requested):
Hi @josh146, good catch with the decomposition! I'm a bit still hesitant with non-tape mode, however.

Would the inverse of the decomposition not be accounted for in the QNode? :thinking: 
https://github.com/PennyLaneAI/pennylane/blob/c0d269e5f1eba2f9d033bd9b6a79c10a11f4228a/pennylane/qnodes/base.py#L100

When decomposing the inverse of the `S` or `T` gates, by taking the inverse of the decomposition on the operation level, it seems that we end up taking the inverse twice overall (and simply applying `S` or `T` eventually):
```python
import pennylane as qml

qml.disable_tape()
dev = qml.device('default.qubit.autograd', wires=1)

if "S" in dev.operations:
    dev.operations.remove("S")

@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.S(wires=0).inv()
    return qml.expval(qml.PauliZ(0))

circuit()
dev.state
```
```
tensor([7.07106781e-01+0.j        , 4.32978028e-17+0.70710678j], requires_grad=True)
```
The same example of using the decomposition in tape mode:
```
tensor([7.07106781e-01+0.j        , 4.32978028e-17-0.70710678j], requires_grad=True)
```
Perhaps adding such examples (or similar) as higher-level integration tests could help here.

josh146 commented:
@antalszava you're entirely right, thank you for noticing that! Considering your comment, I've done the following:

* Removed the inverse branches in the S and T decompositions,

* Modified `Operation.expand()` so that if the original operation is inverted, so is the final decomposition,

* Added a high-level integration test to check all combinations of tape/non-tape mode, and decompositions/non-decompositions.

antalszava commented on tests/devices/test_default_qubit.py:
```diff
@@ -1880,3 +1880,77 @@ def test_partial_subsystem(self, mocker):
+        works when the inverse S gate is decomposed"""
+        dev = qml.device('default.qubit', wires=1)
+
+        patched_operations = dev.operations.copy()
+        patched_operations.remove("S")
+        monkeypatch.setattr(dev, "operations", patched_operations)
```
:100:

antalszava reviewed (approved):
Looks good to me! :1st_place_medal: Great fix :slightly_smiling_face: Was wondering about a test case for `T` too, probably the `S` test case sort of covers the specific case though (as `T` and `S` are very similar gates and the operations themselves are tested separately).
